temporal-sdk/src/main/java/io/temporal/internal/sync/QueryDispatcher.java

```diff
@@ -0,0 +1,111 @@
+ */
+
+package io.temporal.internal.sync;
+
+import io.temporal.api.common.v1.Payloads;
+import io.temporal.common.converter.DataConverter;
+import io.temporal.common.converter.EncodedValues;
+import io.temporal.common.interceptors.WorkflowInboundCallsInterceptor;
+import io.temporal.common.interceptors.WorkflowOutboundCallsInterceptor;
+import io.temporal.workflow.DynamicQueryHandler;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class QueryDispatcher {
+
+  private static final Logger log = LoggerFactory.getLogger(QueryDispatcher.class);
+
+  private WorkflowInboundCallsInterceptor inboundCallsInterceptor;
+  private final DataConverter converter;
+
+  private final Map<String, WorkflowOutboundCallsInterceptor.RegisterQueryInput> queryCallbacks =
+      new HashMap<>();
+
+  private DynamicQueryHandler dynamicQueryHandler;
+
+  public QueryDispatcher(DataConverter converter) {
+    this.converter = converter;
+  }
+
+  public void setInboundCallsInterceptor(WorkflowInboundCallsInterceptor inboundCallsInterceptor) {
+    this.inboundCallsInterceptor = inboundCallsInterceptor;
+  }
+
+  /** Called from the interceptor tail */
+  public WorkflowInboundCallsInterceptor.QueryOutput handleInterceptedQuery(
+      WorkflowInboundCallsInterceptor.QueryInput input) {
+    String queryName = input.getQueryName();
+    Object[] args = input.getArguments();
+    WorkflowOutboundCallsInterceptor.RegisterQueryInput handler = queryCallbacks.get(queryName);
+    Object result;
+    if (handler == null) {
+      if (dynamicQueryHandler != null) {
+        result = dynamicQueryHandler.handle(queryName, (EncodedValues) args[0]);
+      } else {
+        throw new IllegalStateException("Unknown query type: " + queryName);
+      }
+    } else {
+      result = handler.getCallback().apply(args);
+    }
+    return new WorkflowInboundCallsInterceptor.QueryOutput(result);
+  }
+
+  public Optional<Payloads> handleQuery(String queryName, Optional<Payloads> input) {
+    WorkflowOutboundCallsInterceptor.RegisterQueryInput handler = queryCallbacks.get(queryName);
+    Object result;
+    if (handler == null) {
+      if (dynamicQueryHandler != null) {
+        result =
+            inboundCallsInterceptor
+                .handleQuery(
+                    new WorkflowInboundCallsInterceptor.QueryInput(
+                        queryName, new Object[] {new EncodedValues(input, converter)}))
+                .getResult();
+      } else {
+        throw new IllegalStateException("Unknown query type: " + queryName);
+      }
+    } else {
+      Object[] args =
+          DataConverter.arrayFromPayloads(
+              converter, input, handler.getArgTypes(), handler.getGenericArgTypes());
+      result =
+          inboundCallsInterceptor
+              .handleQuery(new WorkflowInboundCallsInterceptor.QueryInput(queryName, args))
```

Comment:
nit: Call to `handleQuery` could be lifted out of if branches which would create the `QueryInput`